### ISSUE TITLE
host key mismatch after upgrading module from v5.12.0 to v5.13.0

### ISSUE THREAD

**Issue by jgoodall** [go-git/go-git#1341]
SSH git clone fails after update to v5.13.0. Specifically, the issue comes from `ssh.FixedHostKey(hostKey)` where we try to set the host key callback. The following works fine with `v5.12.0`, but fails with `v5.13.0`; the error is `ssh: handshake failed: ssh: host key mismatch`. So I dont think it is actually a host key mismatch, but something that changed in this module (or a dependency).

```
	hostKey, _, _, _, err := ssh.ParseAuthorizedKey([]byte(serverHostKey))
	if err != nil {
		return "", fmt.Errorf("unable to parse host key: %w", err)
	}
	pk, err := ssh2.NewPublicKeys(gitUser, []byte(repoPrivateKey), "")
	if err != nil {
		return "", fmt.Errorf("unable to create git public key: %w", err)
	}
	pk.HostKeyCallback = ssh.FixedHostKey(hostKey)
	deployBranch := gitplum.NewBranchReferenceName(gitBranch)
	r, err := git.PlainClone(destDirPath, false, &git.CloneOptions{
		URL:           repoURL,
		Auth:          pk,
	})
	if err != nil {
		return "", fmt.Errorf("unable to clone git repo [%s] (branch [%s]): %w", repoURL, deployBranch, err)
	}
```

I can use insecure and it works, but obviously I would like to verify the host key:

```
	pk.HostKeyCallback = ssh.InsecureIgnoreHostKey()
```

**Comment by pjbgf** [comment]
@jgoodall can you provide some more insights into this? I tried reproducing it with `v5.13.0` against `github.com` and the clone operation worked fine.

The pubkey I used was: `github.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl`.

**Comment by jgoodall** [comment]
It is a private gitlab server. I only bumped go-git (`go get -u github.com/go-git/go-git/v5`), although that did update some required modules:

```
diff --git a/go.mod b/go.mod
index dc8f2d4..e5e229a 100644
--- a/go.mod
+++ b/go.mod
@@ -12 +12 @@ require (
-	github.com/go-git/go-git/v5 v5.12.0
+	github.com/go-git/go-git/v5 v5.13.0
@@ -32 +32 @@ require (
-	github.com/ProtonMail/go-crypto v1.0.0 // indirect
+	github.com/ProtonMail/go-crypto v1.1.3 // indirect
@@ -36 +36 @@ require (
-	github.com/cyphar/filepath-securejoin v0.3.4 // indirect
+	github.com/cyphar/filepath-securejoin v0.3.6 // indirect
@@ -41 +41 @@ require (
-	github.com/go-git/go-billy/v5 v5.5.0 // indirect
+	github.com/go-git/go-billy/v5 v5.6.1 // indirect
@@ -44 +44 @@ require (
-	github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da // indirect
+	github.com/golang/groupcache v0.0.0-20241129210726-2c02b8208cf8 // indirect
@@ -49,0 +50 @@ require (
+	github.com/mmcloughlin/avo v0.6.0 // indirect
@@ -51,2 +52 @@ require (
-	github.com/onsi/gomega v1.33.1 // indirect
-	github.com/pjbgf/sha1cd v0.3.0 // indirect
+	github.com/pjbgf/sha1cd v0.3.1 // indirect
@@ -56 +55,0 @@ require (
-	github.com/rogpeppe/go-internal v1.12.0 // indirect
@@ -67 +66,2 @@ require (
-	golang.org/x/net v0.30.0 // indirect
+	golang.org/x/mod v0.22.0 // indirect
+	golang.org/x/net v0.33.0 // indirect
@@ -68,0 +69 @@ require (
+	golang.org/x/tools v0.28.0 // indirect

```

**Comment by jgoodall** [comment]
Here is code snippet where I am getting the error (constants are environment specific so removed those):


```
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/go-git/go-git/v5"
	gitplum "github.com/go-git/go-git/v5/plumbing"
	ssh2 "github.com/go-git/go-git/v5/plumbing/transport/ssh"
	"golang.org/x/crypto/ssh"
)

const repoHostKey = `<repo-host-key>`
const repoURL = "<repo-url>"
const repoPrivateKey = `<repo-private-key>`

func main() {
	destDirPath := filepath.Join(os.TempDir(), "gittest")
	_ = os.RemoveAll(destDirPath)
	err := os.MkdirAll(destDirPath, 0755)
	if err != nil {
		fmt.Printf("unable to create new git directory [%s]: %s", destDirPath, err)
		return
	}
	hash, err := cloneRepo(repoURL, repoHostKey, repoPrivateKey, destDirPath)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Printf("Cloned repo [%s] to [%s], with head [%s]", repoURL, destDirPath, hash)
	}
}

func cloneRepo(url, serverHostKey, privateKey, dir string) (hash string, err error) {
	hostKey, _, _, _, err := ssh.ParseAuthorizedKey([]byte(serverHostKey))
	if err != nil {
		return "", fmt.Errorf("unable to parse host key: %s", err)
	}
	pk, err := ssh2.NewPublicKeys("git", []byte(privateKey), "")
	if err != nil {
		return "", fmt.Errorf("unable to create git public key: %s", err)
	}
	pk.HostKeyCallback = ssh.FixedHostKey(hostKey)
	deployBranch := gitplum.NewBranchReferenceName("main")
	r, err := git.PlainClone(dir, false, &git.CloneOptions{
		URL:           url,
		Auth:          pk,
		ReferenceName: deployBranch,
		SingleBranch:  true,
		Depth:         1,
	})
	if err != nil {
		return "", fmt.Errorf("unable to clone git repo [%s]: %s", url, err)
	}
	head, err := r.Head()
	if err != nil {
		return "", fmt.Errorf("unable to get git repo [%s] head reference: %s", url, err)
	}
	return head.Hash().String(), nil
}
```

I think it is related to this commit - 20b556be6a825f0255a71376a2bce4c2c8393b89. If I get the commit before this – `go get -u github.com/go-git/go-git/v5@3a3f96c372b0f9a93ea27928baa4516e95a79bde` – it works. But `go get -u github.com/go-git/go-git/v5@20b556be6a825f0255a71376a2bce4c2c8393b89` and it fails.

**Comment by pjbgf** [comment]
@jgoodall I had another go at this, and against GitLab, it worked fine when a `ssh-ed25519` host public key was being provided, but it failed for`ssh-rsa` and `ecdsa-sha2-nistp256`. Can you please confirm what is the Host Key Algorithm that you are experiencing this with?

If worse comes to worse we may have to revert that change and release a patch.
As a workaround, it is possible to create a custom AuthMethod and set `HostKeyAlgorithms` as needed:

```golang
type CustomCallback struct {
	Callback func() (signers []ssh.Signer, err error)
	ssh.HostKeyCallback
}
...
func (a *CustomCallback) ClientConfig() (*ssh.ClientConfig, error) {
	return &ssh.ClientConfig{
		User: "git",
		Auth: []ssh.AuthMethod{ssh.PublicKeysCallback(func() (signers []ssh.Signer, err error) {
			signer, err := ssh2.NewPublicKeys("git", []byte(key), "")
			if err != nil {
				return nil, err
			}

			return []ssh.Signer{signer.Signer}, nil
		})},		
		HostKeyAlgorithms: []string{"ssh-rsa"}, // Use the type based on the hostKey provided. E.g. "ssh-rsa", "ssh-ed25519", "ecdsa-sha2-nistp256"
		HostKeyCallback:   ssh.FixedHostKey(hostKey),
	}, nil
}
```

With a custom AuthMethod all the host keys worked fine for GitLab against both go-git versions.

**Comment by jgoodall** [comment]
I tried each of the available host keys (`ssh-keyscan <server>` to get the keys). 

With `v0.31.12`:

* `ecdsa-sha2-nistp256` - worked
* `ssh-rsa` - `ssh: host key mismatch`
* `ssh-ed25519` - `ssh: host key mismatch`

With `v0.31.13`:

* `ecdsa-sha2-nistp256` -`ssh: host key mismatch`
* `ssh-rsa` - `ssh: host key mismatch`
* `ssh-ed25519` - worked

This makes sense since I had been using the `ecdsa` type to check the host key before upgrading the library. So it sounds like the workaround for me using `v0.31.13` is to use the `ed25519` type. However, if you were to revert the patch I would have to go back to using `ecdsa` type. 

I am not clear on how to use the custom AuthMethod you describe, but sounds like that is the best option - can you provide a more complete example?

**Comment by pjbgf** [comment]
Thanks for that @jgoodall. I have now released the `v5.13.1` patch release with the changes reverted - as at this point in time for `v5` we are not really keen on forcing users to make changes to upgrade.

As for the example you asked:
```golang
func NewCustomAuthMethod(hostPubKey []byte, key []byte, hostKeyAlgo string) (*PublicKeysCallback, error) {
	hostKey, _, _, _, err := ssh.ParseAuthorizedKey(hostPubKey)
	if err != nil {
		return nil, fmt.Errorf("unable to parse host key: %s", err)
	}

	user := "git"
	signer, err := ssh2.NewPublicKeys(user, key, "")
	if err != nil {
		return nil, err
	}

	return &PublicKeysCallback{
		user: user,
		Callback: func() (signers []ssh.Signer, err error) {
			return []ssh.Signer{signer.Signer}, nil
		},
		signers:           []ssh.Signer{signer.Signer},
		HostKeyCallback:   ssh.FixedHostKey(hostKey),
		hostKeyAlgorithms: []string{hostKeyAlgo},
	}, nil
}

type PublicKeysCallback struct {
	Callback func() (signers []ssh.Signer, err error)
	ssh.HostKeyCallback

	user              string
	signers           []ssh.Signer
	hostKeyAlgorithms []string
}

func (a *PublicKeysCallback) Name() string {
	return "ssh-public-key-callback"
}

func (a *PublicKeysCallback) String() string {
	return fmt.Sprintf("user: %s, name: %s", a.user, a.Name())
}

func (a *PublicKeysCallback) ClientConfig() (*ssh.ClientConfig, error) {
	am := ssh.PublicKeysCallback(func() (signers []ssh.Signer, err error) {
		return a.signers, nil
	})
	return &ssh.ClientConfig{
		User:              a.user,
		Auth:              []ssh.AuthMethod{am},
		HostKeyAlgorithms: a.hostKeyAlgorithms,
		HostKeyCallback:   a.HostKeyCallback,
	}, nil
}
```

Then you can call `NewCustomAuthMethod` instead of `ssh2.NewPublicKeys("git", []byte(privateKey), "")`:
```
pk, _ := NewCustomAuthMethod([]byte(repoHostKey), []byte(repoPrivateKey), "ecdsa-sha2-nistp256")
```

Feel free to close this issue if no additional questions remain.

**Comment by evanelias** [comment]
After eyeballing this quickly, I would suspect the problem might be in the error return in these lines https://github.com/go-git/go-git/commit/20b556be6a825f0255a71376a2bce4c2c8393b89#diff-4b7927140f2bd6fe4030126982cb323cf2b49656e75e7f235ca1fbf0336aadf4R141-R143 . Previously with a FixedHostKey callback but no known_hosts file, this would be a harmless no-op (since skeema/knownhosts can't auto-populate the algorithms from a FixedHostKey callback), but after #1157 it would end up being an early error return which skips the subsequent logic in this function. cc @Javier-varez to get more eyes on that.

The full revert of #1157 in #1346 is disappointing and quite unfortunate, as that revert breaks `@cert-authority` support. And again I'm really unhappy with the internal-orphaned-forking of skeema/knownhosts in #1172 as this makes it infinitely harder for me to maintain my code and help out when problems like this come up downstream.

**Comment by pjbgf** [comment]
> The full revert of https://github.com/go-git/go-git/pull/1157 in https://github.com/go-git/go-git/pull/1346 is disappointing and quite unfortunate, as that revert breaks @cert-authority support.

Unfortunately, the `go-git` project have very limited maintainership capacity at the moment, which is mostly focused on v6. Which is the reason why we reverted the changes - lack of time to investigate and fix regressions. We could add the changes back to v5 if someone is willing to fix the issue and propose a new PR, while also be keen on fixing arising issues. 

> I'm really unhappy with the internal-orphaned-forking of skeema/knownhosts in https://github.com/go-git/go-git/pull/1172 as this makes it infinitely harder for me to maintain my code and help out when problems like this come up downstream.

@evanelias As explained on a [separate thread](https://github.com/go-git/go-git/pull/1172#issuecomment-2623793867), those changes are orthogonal as they exist only on the [development branch](https://github.com/go-git/go-git/blob/main/CONTRIBUTING.md#branches). They aim at enabling users to debug their own SSH problems, as per upstream Git behaviour. The `releases/v5.x` branch, which is used to maintain the `v5` release, was not affected by it.

**Comment by evanelias** [comment]
I can completely understand the limited maintainership capacity, and the need to revert problematic PRs. However, it does not appear that anyone tried contacting the original PR author (@Javier-varez) at any point to report there was a problem.

**Comment by pjbgf** [comment]
In the name of transparency, the reason behind the prompt revert was that the `v5.13.0` release contained fixes for the two CVEs in the `go-git` codebase https://github.com/go-git/go-git/security/advisories/GHSA-r9px-m959-cxf4 and https://github.com/go-git/go-git/security/advisories/GHSA-v725-9546-7q7m.

Given the duty of care we have with the `go-git` community, I took the quick decision to revert the change in the belief that it was more important to have a stable patch release for the entire community to upgrade to, than to delay it in the name of a feature which wasn't previously released - and therefore only a subset of users relied on.

The revert is now done and this issue resolve.
